[PyTorch] Narrow Device to 2 bytes by narrowing DeviceType and DeviceIndex

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47023 [PyTorch] Narrow Device to 2 bytes by narrowing DeviceType and DeviceIndex**

DeviceType pretty clearly only needs 1 byte. DeviceIndex only needs 1 byte given that machines don't have anywhere near 255 GPUs in them as far as I know.

Differential Revision: [D24605460](https://our.internmc.facebook.com/intern/diff/D24605460/)

swolchok committed:
[PyTorch] Narrow Device to 2 bytes by narrowing DeviceType and DeviceIndex

DeviceType pretty clearly only needs 1 byte. DeviceIndex only needs 1 byte given that machines don't have anywhere near 255 GPUs in them as far as I know.

Differential Revision: [D24605460](https://our.internmc.facebook.com/intern/diff/D24605460/)

[ghstack-poisoned]

swolchok commented:
I'm assured that actually we only run on 1 box at a time, and we can always correct this later if needed.

swolchok committed:
fix some tests, rebase on "[PyTorch] Narrow Device to 2 bytes by narrowing DeviceType and DeviceIndex"


DeviceType pretty clearly only needs 1 byte. DeviceIndex only needs 1 byte given that machines don't have anywhere near 255 GPUs in them as far as I know.

Differential Revision: [D24605460](https://our.internmc.facebook.com/intern/diff/D24605460/)

[ghstack-poisoned]

glaringlee commented:
@swolchok 
This looks good to me. The only thing I want to check is the PROTO_ONLY_FOR_TEST flag. I am not sure whether removing this flag is a BC change. Can you confirm?

glaringlee reviewed (commented):
@swolchok
This looks good to me. The only thing I want to check is the PROTO_ONLY_FOR_TEST flag. I am not sure whether removing this flag is a BC breaking change. Can you confirm?

swolchok commented:
@glaringlee what does BC stand for?

glaringlee commented:
> @glaringlee what does BC stand for?

Backward compatibility. I am not sure whether PROTO_ONLY_FOR_TEST is used somewhere externally. I saw you removed all the internal usage.

swolchok commented:
> > @glaringlee what does BC stand for?
> 
> Backward compatibility. I am not sure whether PROTO_ONLY_FOR_TEST is used somewhere externally. I saw you removed all the internal usage.

@dzhulgakov tells me that it's some legacy caffe2 thing and we can totally remove it. (right?)

swolchok committed:
fix test_torch.py on "[PyTorch] Narrow Device to 2 bytes by narrowing DeviceType and DeviceIndex"


DeviceType pretty clearly only needs 1 byte. DeviceIndex only needs 1 byte given that machines don't have anywhere near 255 GPUs in them as far as I know.

Differential Revision: [D24605460](https://our.internmc.facebook.com/intern/diff/D24605460/)

[ghstack-poisoned]

swolchok committed:
cast index to int to avoid formatting as char on some platforms on "[PyTorch] Narrow Device to 2 bytes by narrowing DeviceType and DeviceIndex"


DeviceType pretty clearly only needs 1 byte. DeviceIndex only needs 1 byte given that machines don't have anywhere near 255 GPUs in them as far as I know.

Differential Revision: [D24605460](https://our.internmc.facebook.com/intern/diff/D24605460/)

[ghstack-poisoned]

glaringlee commented:
> > > @glaringlee what does BC stand for?
> > 
> > 
> > Backward compatibility. I am not sure whether PROTO_ONLY_FOR_TEST is used somewhere externally. I saw you removed all the internal usage.
> 
> @dzhulgakov tells me that it's some legacy caffe2 thing and we can totally remove it. (right?)

@swolchok Ah, I see, I am lacking of this context. I think this is fine then.

swolchok committed:
rebase on "[PyTorch] Narrow Device to 2 bytes by narrowing DeviceType and DeviceIndex"


DeviceType pretty clearly only needs 1 byte. DeviceIndex only needs 1 byte given that machines don't have anywhere near 255 GPUs in them as far as I know.

Differential Revision: [D24605460](https://our.internmc.facebook.com/intern/diff/D24605460/)

[ghstack-poisoned]

dzhulgakov commented on c10/cuda/CUDAFunctions.cpp:
```diff
@@ -93,6 +93,8 @@ DeviceIndex device_count() noexcept {
   static int count = []() {
     try {
+      auto result = device_count_impl();
+      TORCH_INTERNAL_ASSERT(result < (1 << (sizeof(DeviceIndex) * CHAR_BIT)), "Too many CUDA devices, DeviceIndex overflowed");
```
technically it's a signed value, so you'd need to check 127? you can probably do `std::numeric_limits<DeviceIndex>::max` to avoid the bit shifts

dzhulgakov reviewed (approved):
Looking good!

Some of the CI failures are there but they are all upstream (for future you can try rebasing on viable branch)

As for ONLY_FOR_TEST - yes it's inherited from Caffe2 and it's totally safe to nuke. In fact it's not even used in any tests anymore :)

swolchok committed:
use numeric_limits on "[PyTorch] Narrow Device to 2 bytes by narrowing DeviceType and DeviceIndex"


DeviceType pretty clearly only needs 1 byte. DeviceIndex only needs 1 byte given that machines don't have anywhere near 255 GPUs in them as far as I know.

Differential Revision: [D24605460](https://our.internmc.facebook.com/intern/diff/D24605460/)

[ghstack-poisoned]

swolchok commented:
> Job pytorch_xla_linux_bionic_py3_6_clang9_test is failing. Please create an issue with title prefixed by [PT_BREAK] in pytorch/xla and link to to this PR. If you have questions, please reach out to @ailzhang / @dlibenzi / @JackCaoG.

Am I supposed to land this anyway, or wait for activity on the issue before landing?

swolchok committed:
Update on "[PyTorch] Narrow Device to 2 bytes by narrowing DeviceType and DeviceIndex"


DeviceType pretty clearly only needs 1 byte. DeviceIndex only needs 1 byte given that machines don't have anywhere near 255 GPUs in them as far as I know.

Differential Revision: [D24605460](https://our.internmc.facebook.com/intern/diff/D24605460/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Narrow Device to 2 bytes by narrowing DeviceType and DeviceIndex"


DeviceType pretty clearly only needs 1 byte. DeviceIndex only needs 1 byte given that machines don't have anywhere near 255 GPUs in them as far as I know.

Differential Revision: [D24605460](https://our.internmc.facebook.com/intern/diff/D24605460/)

[ghstack-poisoned]
